stdlib/public/Concurrency/DispatchGlobalExecutor.inc

```diff
@@ -226,6 +224,74 @@ static void swift_task_enqueueGlobalWithDelayImpl(JobDelay delay,
 }
 
 #define DISPATCH_UP_OR_MONOTONIC_TIME_MASK  (1ULL << 63)
+#define DISPATCH_WALLTIME_MASK  (1ULL << 62)
+#define DISPATCH_TIME_MAX_VALUE (DISPATCH_WALLTIME_MASK - 1)
+
+struct __swift_job_source {
+  dispatch_source_t source;
+  Job *job;
+};
+
+static void _swift_run_job_leeway(struct __swift_job_source *jobSource) {
+  dispatch_source_t source = jobSource->source;
+  Job *job = jobSource->job;
+  auto task = dyn_cast<AsyncTask>(job);
+  assert(task && "provided job must be a task");
+  _swift_task_dealloc_specific(task, jobSource);
+  __swift_run_job(job);
+  dispatch_release(source);
+}
+
+#if defined(__i386__) || defined(__x86_64__) || !defined(__APPLE__)
+#define TIME_UNIT_USES_NANOSECONDS 1
+#else
+#define TIME_UNIT_USES_NANOSECONDS 0
+#endif
+
+#if TIME_UNIT_USES_NANOSECONDS
+// x86 currently implements mach time in nanoseconds
+// this is NOT likely to change
+static inline uint64_t
+time_nano2mach(uint64_t nsec) {
+  return nsec;
+}
+#else
+#define DISPATCH_USE_HOST_TIME 1
+#if defined(__APPLE__)
+#if defined(__arm__) || defined(__arm64__)
+// Apple arm platforms currently use a fixed mach timebase of 125/3 (24 MHz)
+static inline uint64_t
+time_nano2mach(uint64_t nsec) {
+  if (!nsec) {
+    return nsec;
+  }
+  if (nsec >= (uint64_t)INT64_MAX) {
+    return INT64_MAX;
+  }
+  if (nsec >= UINT64_MAX / 3ull) {
+    return (nsec / 125ull) * 3ull;
+  } else {
+    return (nsec * 3ull) / 125ull;
+  }
+}
+#endif
+#endif
+#endif
+
+static inline dispatch_time_t
+clock_and_value_to_time(int clock, long long deadline) {
+  uint64_t value = time_nano2mach((uint64_t)deadline);
```

Comment:
Is this an Apple-specific API?  Do we need something else to work with open-source Dispatch?

Reply:
this is partially lifted from the internals of dispatch (such that we produce time values that are compatible with dispatch). the naming is a an adjusted form that mimics the existing dispatch code (which held the mach based names). In non-apple-oses it will just return the value un-modified.

Reply:
renamed it to make a bit more sense.